File: lib/domain/crop/interactors/crop_list_reference_entities_interactor.rb
# frozen_string_literal: true

module Domain
  module Crop
    module Interactors
      class CropListReferenceEntitiesInteractor
        def initialize(output_port:, user_id: nil, gateway:, logger:)
          @output_port = output_port
          @gateway = gateway
          @logger = logger
          @user_id = user_id
        end

        def call(region: nil)
          crops = @gateway.list_by_is_reference(is_reference: true, region: region)
          @output_port.on_success(crops)
        rescue Domain::Shared::Exceptions::RecordInvalid => e
          @output_port.on_failure(Domain::Shared::Dtos::Error.new(e.message))
        end
      end
    end
  end
end

File: crates/agrr-domain/src/crop/interactors/crop_list_reference_entities_interactor.rs
//! Ruby: `Domain::Crop::Interactors::CropListReferenceEntitiesInteractor`

use crate::crop::entities::CropEntity;
use crate::crop::gateways::CropGateway;
use crate::crop::ports::CropListReferenceEntitiesOutputPort;
use crate::shared::dtos::Error;
use crate::shared::exceptions::RecordInvalidError;

pub struct CropListReferenceEntitiesInteractor<'a, G, O> {
    output_port: &'a mut O,
    gateway: &'a G,
}

impl<'a, G, O> CropListReferenceEntitiesInteractor<'a, G, O>
where
    G: CropGateway,
    O: CropListReferenceEntitiesOutputPort,
{
    pub fn new(output_port: &'a mut O, gateway: &'a G) -> Self {
        Self {
            output_port,
            gateway,
        }
    }

    pub fn call(
        &mut self,
        region: Option<&str>,
    ) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
        match self.gateway.list_by_is_reference(true, region) {
            Ok(crops) => {
                self.output_port.on_success(crops);
                Ok(())
            }
            Err(err) => match err.downcast::<RecordInvalidError>() {
                Ok(record_invalid) => {
                    let message = record_invalid
                        .detail_message()
                        .unwrap_or("record invalid")
                        .to_string();
                    self.output_port.on_failure(Error::new(message));
                    Ok(())
                }
                Err(err) => Err(err),
            },
        }
    }
}

#[cfg(test)]
mod interactors_crop_list_reference_entities_interactor_test_inline {
    use super::*;
    include!(concat!(env!("CARGO_MANIFEST_DIR"), "/test/crop/interactors_crop_list_reference_entities_interactor_test.rs"));
}
